WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAddParameterViewModel.swift

```diff
@@ -0,0 +1,76 @@
+import Foundation
+
+/// View model for `BlazeAddParameterView`
+final class BlazeAddParameterViewModel: ObservableObject {
+    @Published var key: String
+    @Published var value: String
+    @Published var hasValidationError: Bool = false
+    @Published var hasCountError: Bool = false
+
+    let remainingCharacters: Int
+    let isNotFirstParameter: Bool
+    let parameter: BlazeAdURLParameter?
+
+    // For adding or editing a new parameter, the two inputs are to be combined to be "key=value".
+    // However, for adding or editing 2nd or more parameters, the input becomes "&key=value" due to how URL parameter work.
+    var totalInputLength: Int {
+        let inputString = (isNotFirstParameter ? "&" : "") + key + "=" + value
+        return inputString.count
+    }
+
+    var shouldDisableSaveButton: Bool {
+        key.isEmpty || value.isEmpty || hasCountError || hasValidationError
+    }
+
+    typealias BlazeAddParameterCompletionHandler = (_ key: String, _ value: String) -> Void
+    private let completionHandler: BlazeAddParameterCompletionHandler
+
+    private let cancellationHandler: () -> Void
+
+    init(remainingCharacters: Int,
+         isNotFirstParameter: Bool = true,
+         parameter: BlazeAdURLParameter? = nil,
+         onCancel: @escaping () -> Void,
+         onCompletion: @escaping BlazeAddParameterCompletionHandler) {
+        self.remainingCharacters = remainingCharacters
+        self.isNotFirstParameter = isNotFirstParameter
+        self.parameter = parameter
+        self.cancellationHandler = onCancel
+        self.completionHandler = onCompletion
+
+        key = parameter?.key ?? ""
+        value = parameter?.value ?? ""
+    }
+
+    func didTapCancel() {
+        cancellationHandler()
+    }
+
+    func didTapSave() {
+        completionHandler(key, value)
+    }
+
+    func validateInput(text: String) {
+        validateParameter(text: text)
+        validateInputLength()
+    }
+
+    /// This function validates the URL parameters using String.isValidURL().
+    /// It requires a full URL, thus Constant.baseURLForValidation is added.
+    /// The constant uses "?key=", but it's also used to validate parameter keys, as they both follow the same validation rule.
+    ///
+    private func validateParameter(text: String) {
+        let url = Constant.baseURLForValidation + text
+        hasValidationError = !url.isValidURL()
+    }
+
+    private func validateInputLength() {
+        hasCountError = remainingCharacters - totalInputLength <= 0
```

Comment:
Nit: since `totalInputLength` is not reused any where else, please keep it in this method for readability and private scope.

Reply:
Nice, thanks! At first I had in mind to display total length on the UI somewhere, then cancelled the idea but forgot to refactor.

Added in https://github.com/woocommerce/woocommerce-ios/pull/11642/commits/b3b72c4ce4936a8280dd755888aef3e9b5867c62